### ISSUE TITLE
Runtime bug: Cycling of weapons not as expected

### ISSUE THREAD

**Issue by InterClaw** [dxx-rebirth/dxx-rebirth#757]
#### Operating System Environment

* Microsoft Windows
  * [ ] Microsoft Windows XP
  * [ ] Microsoft Windows Vista
  * [ ] Microsoft Windows 7
  * [ ] Microsoft Windows 8
  * [ ] Microsoft Windows 8.1
  * [x] Microsoft Windows 10

* Mac OS X
  * [ ] Mac OS X 10.10
  * [ ] Mac OS X 10.11
  * [ ] Mac OS X 10.12
  * [ ] Mac OS X 10.13
  * [ ] Mac OS X 10.14
  * [ ] Mac OS X 10.15

* Debian
  * [ ] Debian Stretch
  * [ ] Debian Buster
  * [ ] Debian Bullseye
  * [ ] Debian Sid
* Fedora
  * [ ] Fedora 28
  * [ ] Fedora 29
  * [ ] Fedora 30
  * [ ] Fedora 31
  * [ ] Fedora 32
  * [ ] Rawhide
* Ubuntu
  * [ ] Ubuntu 16.04 LTS (Xenial Xerus)
  * [ ] Ubuntu 18.04 LTS (Bionic Beaver)
  * [ ] Ubuntu 18.10 (Cosmic Cuttlefish)
  * [ ] Ubuntu 19.04 (Disco Dingo)
  * [ ] Ubuntu 19.10 (Eoan Ermine)
  * [ ] Ubuntu 20.04 LTS (Focal Fossa)

* [ ] Arch
* [ ] Gentoo

* [ ] OpenBSD
* [ ] FreeBSD
* [ ] NetBSD

#### CPU environment

* [ ] x86 (32-bit Intel/AMD)
* [x] x86\_64 (64-bit Intel/AMD)
* [ ] ARM (32-bit)
* [ ] ARM64 (64-bit; sometimes called AArch64)

### Description

This MIGHT be me not fully understanding the intended behavior, but here goes...

I have "Only cycle autoselect weapons" enabled.

If I select a weapon that is set to "Never autoselect below" via a normal weapon key, the cycle weapon button first cycles through the remaining weapons of the "Never autoselect below" list, before it enters into the list of weapons it SHOULD cycle. After that it continues to only cycle the weapons it should cycle.

That might be a bit hard to follow, so let me explain some more.

What I want to accomplish is the following:
- I want a dedicated button to select the "Laser cannon" and then I want the "Cycle primary" button to cycle the other, more "exotic" primary weapons, except some weapons that I never want to use (e.g. the "Spreadfire cannon").
- Similarly I want a dedicated button for the "Concussion missile" and the "Cycle secondary" button to cycle the other missiles, starting with the important "Homing missile". Except bombs, since I have the bombs on another, dedicated button, so no need to cycle them.
- (In Descent 2, the dedicated buttons actually cycle two weapons, but that's acceptable.)
- The intention here is of course to be able to play the game with only a standard controller, so there are a limited number of buttons available to handle a lot of weapons. I use D-pad Down/Right for laser/cycle and A(Cross)/X(Square) for concussion/cycle.

With the settings below, if I press "1" then "Cycle primary", I don't get the "Vulcan cannon" as expected. Instead I get the "Super laser cannon" and then the "Spreadfire cannon" and only after that the "Vulcan cannon". "Super laser cannon" and "Spreadfire cannon" are the ones I'm trying to avoid. The first I can get with two presses on my dedicated button and the "Spreadfire cannon" is trash anyway, so I always want to skip it.

This can somewhat be worked around by placing the "Laser cannon" and "Concussion missile" at the BOTTOM of the lists of "Never autoselect below", since then the cycle buttons will step to the first listed weapon to cycle. This works fine in Descent 1 to avoid both the "Spreadfire cannon" and the "Proximity bomb".

However, in Descent 2 you also have to place the "Super laser cannon" and "Flash missile" somewhere on the list of "Never autoselect below" and the best you can do then is second to last. So if you have either of those selected, the cycle button will then first have to step to the "Laser cannon" or "Concussion missile" respectively, before going to the intended weapons in the cycle list.

The problem with this is that you can't build up a muscle memory of how many presses of the cycle button it takes to get a certain weapon, because it depends on what you currently have selected. I want to be able to press "1" -> "Cycle primary" -> "Cycle primary" and know I have the "Gauss cannon". Similarly, "Cycle secondary" -> "6" guarantees the "Concussion missile".

I propose that no weapon in the "Never autoselect below" should ever be cycled to if you have "Only cycle autoselect weapons" enabled. You can still use dedicated weapon buttons for any weapon (cyclable or not). But if you press the cycle button it should directly jump to the first weapon on the cycle list.

I'm using this version:
dxx-rebirth_20220929-win.zip

![Screenshot 2024-01-21 233635](https://github.com/dxx-rebirth/dxx-rebirth/assets/48981208/ac930f43-b354-4872-811e-4b8c70c99106)

![Screenshot 2024-01-21 233640](https://github.com/dxx-rebirth/dxx-rebirth/assets/48981208/ff4e3abe-8503-4a72-b2f1-c1d2566c5bc5)

![Screenshot 2024-01-21 234429](https://github.com/dxx-rebirth/dxx-rebirth/assets/48981208/e6a8cd8f-bdc8-44f5-9916-7b81ac7c31c2)

![Screenshot 2024-01-21 234305](https://github.com/dxx-rebirth/dxx-rebirth/assets/48981208/af780921-34cd-48be-88a3-b6ff5e41ba83)

![Screenshot 2024-01-21 233644](https://github.com/dxx-rebirth/dxx-rebirth/assets/48981208/40e513ba-79ce-44bf-80fa-2ab00b3d7d5e)

### Steps to Reproduce

* Engines affected:
  * [x] D1X-Rebirth
  * [x] D2X-Rebirth

* Game modes affected:
  * [x] Single player
  * [ ] Multiplayer competitive
    * [ ] Anarchy
    * [ ] Team Anarchy
    * [ ] Robo-Anarchy
    * [ ] Capture the Flag
    * [ ] Bounty
  * [ ] Multiplayer cooperative

* Players affected by issue:
  * [ ] Game host
  * [ ] Game guests

**Comment by InterClaw** [comment]
I realized that I can mostly attain this by setting it up like below. Then the cycle button will predictably select the top most weapon (if I have one of the dedicated weapons selected).

![Screenshot 2024-01-22 020424](https://github.com/dxx-rebirth/dxx-rebirth/assets/48981208/c4e1e4f7-7886-4c74-bc41-d101dbfdd4c7)

![Screenshot 2024-01-22 020444](https://github.com/dxx-rebirth/dxx-rebirth/assets/48981208/227493ee-abc2-4f21-9b68-2b7679348650)

One downside with this is that the "Super laser vannon" and "Flash missile" are part of the cycle now, which is unnecessary.

I also realized that the dedicated weapon keys will always select the one of the two weapons that was last selected, however that weapon was selected. So if I have the "Concussion missile" selected and then cycle secondaries, if I go past the "Flash missile" and press the dedicated button I will select the "Flash missile" and not the "Concucssion missile", since "Flash missile" was the one of the two that was last selected. Oh well.

I get that this memory function for the dedicated weapon keys is probably holy, but with the mappings above it decreases the chances of being able to go back to your standard weapons somewhat. An option to be able to turn off the memory function of the dedicated weapon keys would be awesome. Then you would always end up on the first of the two weapons when pressing the dedicated key when you have something else selected. Subsequent presses would toggle back and forth between the two as normal. Double press "2" will always give "Gauss cannon" (unless you have the "Vulcan cannon" out). Single press "7" will always give "Homing missile" (unless you already have the "Homing missile"). This would add a lot of predictability. I guess this is a feature request. But it's kind of related to the original bug report.

**Comment by vLKp** [comment]
> I get that this memory function for the dedicated weapon keys is probably holy,

The memory behavior is specifically documented in the manual released with the original game.

If you have a key bound specifically to pick "Vulcan Cannon" and that ever picks Gauss Cannon, that does sound surprising.  Where did you set that?  Your screenshot shows that the regular/super weapon menu is used, and that one does pick between the two.

Cycling reuses the auto-select ordering, because people generally want the same order for both purposes.  Auto-select can never cross the marker line.  An initial read of the old code suggests that at one time cycling would automatically move to the top of the list if the current weapon was below the cutpoint line.  That no longer happens.  It looks like this was changed in 9de6618ba0f5eabbb619044cec5454ac83714aa4, but I have not tested that.

**Comment by InterClaw** [comment]
> Cycling reuses the auto-select ordering, because people generally want the same order for both purposes. Auto-select can never cross the marker line. An initial read of the old code suggests that at one time cycling would automatically move to the top of the list if the current weapon was below the cutpoint line. That no longer happens. It looks like this was changed in [9de6618](https://github.com/dxx-rebirth/dxx-rebirth/commit/9de6618ba0f5eabbb619044cec5454ac83714aa4), but I have not tested that.

Ah, this sounds like that at one time, if you had manually selected a weapon below the cutoff point, if you hit the cycle button, you would immediately jump to the top weapon of the cycle list? I.e. skipping any additional non-cycle weapons below the line? Hm, if I understood you correctly, then this would have been the behavior I wanted...

> If you have a key bound specifically to pick "Vulcan Cannon" and that ever picks Gauss Cannon, that does sound surprising. Where did you set that? Your screenshot shows that the regular/super weapon menu is used, and that one does pick between the two.

But this does sound promising! So there's a way to bind a specific weapon to a button and then ONLY that weapon will ever be selected by that button ever (no toggling between the normal/super weapon)? If so, then that would solve my problem completely I think. I could then bind the "Concussion missile" ONLY to a button and then sort that to the very bottom of the list. Then just put the "Flash missile" somewhere at the top in the cycle list and I can then predictably select any of the secondaries I want there.

So how do you bind just one weapon to a key? I only found this menu in the UI.

![image](https://github.com/dxx-rebirth/dxx-rebirth/assets/48981208/8754dfa3-be79-417f-9336-6bd68e872aec)

Found the `.plx` file and the same mappings. Do I set up a `[weapon keys v1]` section in parallel to this somehow?

```
[weapon keys v2]
1=0x2,0xc,0xff
2=0x3,0xff,0xff
3=0x4,0xff,0xff
4=0x5,0xff,0xff
5=0x6,0xff,0xff
6=0x7,0x0,0xff
7=0x8,0xff,0xff
8=0x9,0xff,0xff
9=0xa,0xff,0xff
0=0xb,0xff,0xff
[end]
```

**Comment by vLKp** [comment]
@InterClaw wrote:

> I have "Only cycle autoselect weapons" enabled.

As I read the current code, the effect of this toggle controls what the game should do when you press cycle, and it encounters the never-below barrier.  When the toggle is clear, it moves to the next real weapon below, and so will always visit every weapon you possess.  When the toggle is set, reaching that barrier forces it to return to the top of the list, so it can never jump from an auto-selectable weapon to a non-auto-selectable weapon.  In neither case will this affect the logic when you start out below the line, as a result of having explicitly selected a weapon that can never be auto-selected.  If you are below the line, it will advance until it either loops back to the top or finds a weapon you possess.

The whole design of reusing the autoselect list to set the sequence for the cycle list seems like a compromise built on the idea that people generally want the same behavior for both.

@InterClaw wrote:
> Ah, this sounds like that at one time, if you had manually selected a weapon below the cutoff point, if you hit the cycle button, you would immediately jump to the top weapon of the cycle list?

That is how I interpret the old code, yes.  The commit message does not explain the rationale for why this was changed.  As I was not the author of that commit, I cannot comment on the author's intention.

> So there's a way to bind a specific weapon to a button and then ONLY that weapon will ever be selected by that button ever (no toggling between the normal/super weapon)?

I am not aware of such a button, but your prior text suggested you had one, which is why I asked where you set that.

No, `weapon keys v1` would be ignored.  The game has no code to recognize such a section.

Regarding predictability, remember that the UI shows you which weapons currently have last-is-super set vs clear, so you can tell from the display whether pressing Vulcan/Gauss will select Vulcan next or Gauss next.

---

This seems to me like you are trying to solve the wrong problem.  Instead of using the cycle button so much, you should bind a keyboard key or joystick button to select the weapon you actually want.  This would also be faster, as it would mean fewer presses to reach your intended destination.

**Comment by InterClaw** [comment]
Ah I guess I confused this. What I meant by the "dedicated" weapon buttons was only the standard weapon buttons that toggle between normal/super.

Yeah, I wonder why the cycle works that way then. Oh well, as mentioned before, I can mostly work around it (in the manner mentioned in my [2nd post](https://github.com/dxx-rebirth/dxx-rebirth/issues/757#issuecomment-1902858206)). Only have to remember what was selected last between laser/super laser and concussion/flash.

Only have two buttons per weapon type to spare for this on the controller. :) So one "shortcut" and one cycle button each. That's it. Don't want to reach for the keyboard.
